Children residence change/file signature

Rikkiabb opened the pull request:
# \<Description\>
Adds two new endpoints to support electronic signature for the application system. 
1. Is requesting a file signature from dokubit which returns a controlCode to display to the user and a documentToken. 
2. Uses the documentToken to see if user has authenticated and if so uploads the signed document to s3.

TODO: Get environment variables from Dokobit and add it to env.

## Why
Needed for children residence change application.


## Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] Formatting passes locally with my changes
- [X] I have rebased against main before asking for a review

VilliGunn commented on apps/application-system/api/src/app/modules/application/files/file.service.spec.ts:
```diff
@@ -1,114 +1,191 @@
+
+  const createApplication = (
+    answers: object = {
+      email: 'email@email.is ',
```
```suggestion
      email: 'email@email.is',
```

Rikkiabb replied:
🙈

VilliGunn commented on apps/application-system/api/src/app/modules/application/files/file.service.spec.ts:
```diff
@@ -1,114 +1,191 @@
+    controlCode: 'code',
+    documentToken: 'token',
+  }
+  const applicantsPhoneNumber = '111-2222'
```
I am wondering if we want to have all the test data in the same place. I get that the data that is used in more than place has been extracted to a constant but in my opinion it would be super nice to have all the data in one place and constructed with an object:

```
const parentA: {
    phoneNumber: '111-2222',
    name: 'Test name',
}
const parentB: {
    phoneNumber: '222-1111',
    name: 'Parent B name',
}
```
etc.

VilliGunn commented on apps/application-system/api/src/app/modules/application/files/utils/aws.ts:
```diff
@@ -0,0 +1,55 @@
+import * as AWS from 'aws-sdk'
+import { environment } from '../../../../../environments'
```
This feels super weird to see in a repo full of aliases. Should we create an alias if there isn't one for this?

Rikkiabb replied:
Totally agree, this is done this way in 35 places in the repo, so I am guessing there is some reason for why there is not an alias used for the environment. Feel that this is out of scope for this PR since this looks like the pattern that is being created for services in BE.

Rikkiabb replied:
Just noticed that this is also fixed in this PR: https://github.com/island-is/island.is/pull/2973/files/9241da83202f9ed51dce3cf43ac0cbaf89e309e2..37882b8f601e6332ae7caf3837df9922186a591e

VilliGunn commented on apps/application-system/api/src/app/modules/application/files/utils/aws.ts:
```diff
@@ -0,0 +1,55 @@
+
+const s3 = new AWS.S3()
+const oneMinute = 60
+const bucket = environment.fsS3Bucket ?? 'development-legal-residence-change'
```
Do we want to hardcode the default here? Don't we want this to fail on prod for example if the env variable is missing?

VilliGunn commented on apps/application-system/api/src/app/modules/application/files/utils/pdf.ts:
```diff
@@ -4,7 +4,7 @@ import {
   PersonResidenceChange,
 } from '@island.is/application/templates/children-residence-change'
-import { Constants } from './constants'
+import { PdfConstants as Constants } from './constants'
```
Why not just use `PdfConstants` instead of renaming to `Constants` ?

VilliGunn reviewed (commented):
Just some minor comments from me 💯

eirikurn commented on apps/application-system/api/src/app/modules/application/files/file.service.ts:
```diff
@@ -40,97 +41,95 @@ export class FileService {
+      .getSignedDocument(DokobitFileName[type], documentToken)
+      .then((file) => {
+        const s3FileName = `${BucketTypePrefix[type]}/${application.id}.pdf`
+        uploadFile(Buffer.from(file, 'binary'), bucket, s3FileName)
```
We should use Buffers instead of binary strings whenever possible.

eirikurn commented on apps/application-system/api/src/app/modules/application/files/file.service.ts:
```diff
@@ -40,97 +41,95 @@ export class FileService {
-      })
+    const s3FileName = `${BucketTypePrefix[type]}/${applicationId}.pdf`
+    const s3File = await getFile(bucket, s3FileName)
+    const fileContent = s3File.Body?.toString('binary')
```
We should use buffers instead of binary strings whenever possible.

Rikkiabb replied:
Yes agree. 
When implementing this I tried that approach and when you rightfully poked at this then I tried at again but no luck 😭 
The problem I think is the way dokobit works. If I send anything else then a binary string to them then I get an error from their side or and empty document. I talked to @gudjong  about this since he implemented the signing service and he had the same experience. If you have any suggestions to try out them I am all ears. 🙏

eirikurn replied:
I'm just talking about it on our end. 

The dokubit nest service is doing `Base64.btoa(string)` before sending the document to their API. So instead of us converting buffers to binary strings to base64, we can just pass buffers around and then do `buffer.toString('base64')` which should at least save one copy of the file in memory, and the base64 encoding probably runs in native code.

Rikkiabb replied:
Yes totally agree on paper. Tried this out at first and then did it again now to make sure I wasn't doing something silly. This is the response I get from Dokobit when I try that: 
`Uploaded document is not compliant with PDF standard` 🤷

eirikurn replied:
Whaaaaaat??? 🤯

eirikurn commented on apps/application-system/api/src/app/modules/application/application.controller.ts:
```diff
@@ -572,4 +574,63 @@ export class ApplicationController {
+      {
+        attachments: {
+          ...application.attachments,
+          fileSignature: {
+            controlCode: controlCode,
+            documentToken: documentToken,
+          },
```
For now, I would rather you store this under `application.externalData.fileSignature`.

eirikurn commented on apps/application-system/api/src/app/modules/application/files/utils/aws.ts:
```diff
@@ -0,0 +1,58 @@
+  await s3
+    .upload(uploadParams)
+    .promise()
+    .catch((err) => {
+      return null
+    })
```
Why swallow errors?

eirikurn commented on apps/application-system/api/src/app/modules/application/files/utils/aws.ts:
```diff
@@ -0,0 +1,58 @@
+    Expires: oneMinute * 120, // TODO: Select length for presigned url's in island.is
+  }
+
+  return await new Promise((resolve, reject) => {
+    s3.getSignedUrl('getObject', presignedUrlParams, (err, url) => {
+      err ? reject(err) : resolve(url)
+    })
+  })
```
Why not getSignedUrlPromise?

eirikurn commented on apps/application-system/api/src/app/modules/application/files/utils/aws.ts:
```diff
@@ -0,0 +1,58 @@
+    Bucket: bucket,
+    Key: fileName,
+  }
+  return await new Promise((resolve, reject) => {
+    s3.getObject(downloadParams, (err, res) => {
+      err ? reject(err) : resolve(res)
+    })
+  })
```
Why not getObject(downloadParams).promise()?

eirikurn commented on apps/application-system/api/src/app/modules/application/files/file.service.spec.ts:
```diff
@@ -1,122 +1,215 @@
       ],
     }).compile()
 
+    jest
+      .spyOn(aws, 'getFile')
+      .mockImplementation(() => Promise.resolve({ Body: 'body' }))
```
Huh. Does this work?

Rikkiabb replied:
Yup do you want to go for another approach?

eirikurn replied:
Just surprised at this Jest black-magic, since I thought this module had a separate "aws" object from what is imported elsewhere.

If it works, then its fine, otherwise I would have suggested overriding it in a Nest TestingModule.

eirikurn commented on apps/application-system/api/src/app/modules/application/files/utils/aws.ts:
```diff
@@ -0,0 +1,58 @@
+import * as AWS from 'aws-sdk'
+
+const s3 = new AWS.S3()
```
Should this be a Nest module/service? To make it easier/cleaner to inject and stub.

Rikkiabb replied:
Fixed ✅  but want to pick your 🧠. Do you have any getlines that you follow when an utility is a service vs a module vs utility functions in the monorepo?

eirikurn replied:
My thinking is if it makes sense to dependency inject it, have it configured externally, or if I want a "simple" way to override/stub it in tests, then we should inject it using NestJS.

Usually applies to stuff that has external dependencies or side effects that we want to configure or stub in tests.

While pure utility functions don't necessarily need to be part of the NestJS DI container.

eirikurn commented on libs/api/domains/application/src/lib/application.resolver.ts:
```diff
@@ -142,4 +144,23 @@ export class ApplicationResolver {
+    )
+  }
+
+  @Mutation(() => Application, { nullable: true })
+  uploadSignedFile(
+    @Args('input') input: UploadSignedFileInput,
+    @CurrentUser() user: User,
+  ): Promise<Application> {
+    return this.applicationService.uploadSignedFile(input, user.authorization)
+  }
```
Where will this mutation be triggered? Shouldn't the user _just_ start the signing process, and when that's done, we somehow upload the file automatically?

Rikkiabb replied:
Yes this flow is weird. So currently it is like this:
1. User requests for file signature. We send back a control code (pin) which is displayed to the user and a document token. 
2. When the user receives these items we send another request which calls the signing service with that document token to check if the signature was successful. In the meantime the user approves/rejects electronic signature on their phone.
It would be ideal that this was one call from the client but since we have to return the control code and then know if the file upload was successful in the client then I implemented this as two calls. This is the same way as it is implemented in the judicial system. 
We could go for server sent event here. But not sure how that works with a graphql layer or if that is a overkill. Any suggestions/thoughts?

eirikurn replied:
Ok, thanks for the explanation. I think we can revisit this later.

eirikurn commented on apps/application-system/api/src/app/modules/application/files/utils/pdf.ts:
```diff
@@ -14,12 +14,12 @@ export async function generateResidenceChangePdf(
+    size: PdfConstants.PAGE_SIZE,
     margins: {
-      top: Constants.VERTICAL_MARGIN,
-      bottom: Constants.VERTICAL_MARGIN,
-      left: Constants.HORIZONTAL_MARGIN,
-      right: Constants.HORIZONTAL_MARGIN,
+      top: PdfConstants.VERTICAL_MARGIN,
+      bottom: PdfConstants.VERTICAL_MARGIN,
+      left: PdfConstants.HORIZONTAL_MARGIN,
+      right: PdfConstants.HORIZONTAL_MARGIN,
     },
   })
```
I dream about having a PDF generation service, which supports multiple types of PDFs, accepts data through a POST body and renders the PDF either with react-pdf or plain react website that gets saved to PDF with the puppeteer.

Maybe one day :)

eirikurn reviewed (commented):
Generally, this could have more guards, since the order of operations is fairly brittle, also make sure it's only used in specific types of applications and signed by the right users.

Going forward, I'd really like to refactor this to an external data + template module system. We'll discuss that later.

gudjong commented on apps/application-system/api/src/app/modules/application/application.module.ts:
```diff
@@ -50,8 +51,13 @@ if (process.env.INIT_SCHEMA === 'true') {
-    UploadProcessor,
     FileService,
+    UploadProcessor,
+    {
+      provide: SIGNING_OPTIONS,
+      useValue: environment.signingOptions,
+    },
+    SigningService,
```
This one is on me and not really in the scope of this PR. I was not aware of the `register` approach when implementing the library. It is on my todo list to change this.

Rikkiabb commented:
> Generally, this could have more guards, since the order of operations is fairly brittle, also make sure it's only used in specific types of applications and signed by the right users.


All for adding guards here. To make sure we are on the same page. You would like to have three guards.
1. Order of operations
2. Only for specific type of applications
3. Signed by the correct users

Or are you thinking a single guard for all of this? And is there something I am missing? 😊

eirikurn commented:
No, that's great. Your call if it makes sense to create Nest guards for this or just add `if` statements in the right places.

Rikkiabb commented:
I have added guards for order of operations and application types. 
I added if statements because I don't have access to the application as a whole in the guards unless I would look it up in the db using the id and I don't want to do a double db look up for each api call on these endpoints. 
I had some trouble figuring out how I would be able to check who are the right users to sign. I currently have who owns the application and then parent B will be an assignee, but I would have to validate those ssn's against something to check if they have access to sign the files, but not sure what this something should be 🤷  Am I way off here or were you thinking something else @eirikurn?

eirikurn commented on apps/application-system/api/src/app/modules/application/application.controller.ts:
```diff
@@ -558,6 +560,8 @@ export class ApplicationController {
   ): Promise<ApplicationResponseDto> {
     const { type } = input
 
+    this.fileService.validateApplicationType(application.typeId)
```
Why not put the guards inside the corresponding functions? 😶😊

Rikkiabb replied:
I actually started that way and then I thought about who should have this responsibility and decided on this should be on a higher level. Up for discussion imo 😄

eirikurn reviewed (approved):
Sweet! LGTM. Just minor comments that you may address in later PRs if you want.

Rikkiabb commented on apps/application-system/api/src/app/modules/application/files/file.service.ts:
```diff
@@ -40,97 +46,141 @@ export class FileService {
-      ContentDisposition: 'inline',
-      ContentType: 'application/pdf',
-      Body: buffer,
+  validateApplicationType(applicationType: string) {
```
This is a very valid point, my feeling is iterate when we have the second pdf contract. Since then we have a more mature vision of this.
My current view of this and I hoped that this reflected that each applicaiton can only create a single pdf and then the pdf type enum should only have values that are reflected in the application type enum and you need to add to the pdf enum to be able to have a pdf from that applicaiton. 
I feel that it will be very valuable to get your input once the next pdf is created and if we will create that then I will make sure you are requested for that review.
Resolving this for now and we will iterate once we get there :)

VilliGunn reviewed (approved):
Nothing further from me after this great thorough review from @eirikurn 💯
